src/main/java/com/alvarium/contracts/RequestSpecialityComponent.java

```diff
@@ -0,0 +1,46 @@
+
+/*******************************************************************************
+ * Copyright 2022 Dell Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package com.alvarium.contracts;
+
+public enum RequestSpecialityComponent {
+  METHOD("@method"),
+  TARGETURI("@target-uri"),
+  AUTHORITY("@authority"),
+  SCHEME("@scheme"),
+  PATH("@path"),
+  QUERY("@query"),
+  QUERYPARAMS("@query-params");
+
+  private final String value;
+
+  private RequestSpecialityComponent(String value) {
+    this.value = value;
+  }
+
+  public String getValue() {
+    return this.value;
+  }
+
+  public static RequestSpecialityComponent fromString(String value) {
+    for (RequestSpecialityComponent e : RequestSpecialityComponent.values()) {
+      if (e.value.equals(value)) {
+        return e;
+      }
+    }
+    return null;
```

Comment:
I have a few comments here:

1. It seems like this should return [EnumConstantNotPresentException](https://docs.oracle.com/javase/7/docs/api/java/lang/EnumConstantNotPresentException.html). This would allow you to use a try/catch in ParseResult.java rather than checking for null. `NULL` has no semantic meaning whereas a typed Exception does. The developer who comes after you will have to dig into the `fromString()` implementation to understand what `NULL` means but you can easily communicate what it means and save them the trouble via a typed Exception.

2.  It's unfortunate to execute a loop to return a value every time a string is passed in. This section of code is going to be called **a lot**. How can the lookup be made more efficient given that all elements in the enum must be unique?

3.) By convention the variable `e` is often a short-hand for Exception in try/catch blocks. Use something else...like `c` or `x` or anything else.

Reply:
Fixed in [dcd31ee](https://github.com/project-alvarium/alvarium-sdk-java/commit/dcd31ee4a3869875e07ccf0c28a96420e26ec365)

To make the lookup more efficient we added `static Map<String, DerivedComponent> derivedComponentMap` in `DerivedComponent` Enum. `derivedComponentMap` is initialized in static initialization block. We can then use `derivedComponentMap` in `fromString()` to return the `DerivedComponent` that maps to the given `value` in 1 lookup.